Fix panic when parsing vergen version

klochowicz opened the pull request:
Vergen version was used to determine need for heartbeats.
Whilst at there, move the version without the heartbeats to past 0.5

klochowicz committed:
Do not panic when cannot parse vergen version

We should never `expect()` on input that we don't fully control.

Instead of panicking, return an anyhow::error, and keep sending legacy
heartbeats just in case.

bonomat commented on maker/src/connection.rs:
```diff
@@ -39,13 +39,18 @@ use xtras::SendInterval;
-    let vergen = semver::Version::parse(vergen_string).expect("daemon version to parse");
+fn parse_vergen_version(vergen_string: &str) -> Result<semver::Version> {
+    let vergen = semver::Version::parse(vergen_string)
+        .context("Could not parse daemon_version: {vergen_string}")?;
```
Do you know what the version was when it paniced?

klochowicz replied:
unfortunately, it printed only the offending character, which was, if I remember correctly `<` (None that this would not have happened in production, as we only make weird tags where we append git hashes when building from `master`.

klochowicz replied:
I have reworked the PR slightly to actually log the error when we encounter offending string, so we have more information now.

luckysori commented on maker/src/connection.rs:
```diff
@@ -39,13 +39,18 @@ use xtras::SendInterval;
-    let vergen = semver::Version::parse(vergen_string).expect("daemon version to parse");
+fn parse_vergen_version(vergen_string: &str) -> Result<semver::Version> {
+    let vergen = semver::Version::parse(vergen_string)
+        .context("Could not parse daemon_version: {vergen_string}")?;
```
🔧 Ran into this myself recently, it's easy to miss:

```suggestion
        .with_context(|| format!("Could not parse daemon_version: {vergen_string}"))?;
```

Otherwise you will print the static string `"{vergen_string}"`.

bonomat replied:
haha, good find 😅

klochowicz replied:
omg, that is indeed easy to miss!

luckysori commented on maker/src/connection.rs:
```diff
@@ -39,13 +39,18 @@ use xtras::SendInterval;
-    let vergen = semver::Version::parse(vergen_string).expect("daemon version to parse");
+fn parse_vergen_version(vergen_string: &str) -> Result<semver::Version> {
+    let vergen = semver::Version::parse(vergen_string)
+        .context("Could not parse daemon_version: {vergen_string}")?;
```
🙃 

It's weird that the context talks about `daemon` here, when this is just a helper that maps string to `semver::Version`.

klochowicz replied:
maybe I'll rename the helper to `parse_daemon_version`? I could create a newtype, but I'm not sure whether this would help us much in this case!

luckysori replied:
It's very nitpicky (hence the 🙃), but it would make more sense to me to:

- Keep the helper as a general helper called `parse_vergen_version`.
- Remove the `context` and make the helper return a `Result<semver::Version, semver::Error>`.
- Add the `context` to line 503, where the helper is called (or alternatively to the trace on line 510).

klochowicz replied:
I'm sorry, I completely missed this comment, only noticed now when going through my notifications! I'll put it on my list of small fixes.

luckysori replied:
I think what you did is fine too! Just a tiny style thing almost.

klochowicz committed:
Do not remove heartbeats until after v0.5

We are still using legacy networking connection; do not remove heartbeats until
we support it.
